{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pysam\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_dir = '/data/parastou/RNAdeg/results/RipChip/bams/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir = '/data/parastou/RNAdeg/CoveragePlots/ChIP/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_coverage_dict(coverage_dict, ch, start, end, inc):\n",
    "    \n",
    "    for i in range(start, end):\n",
    "        key = ch + ':' + str(i)\n",
    "        if key in coverage_dict:\n",
    "            coverage_dict[key] += inc\n",
    "        else:\n",
    "            coverage_dict[key] = inc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_coverage_dict(alignmentFile):\n",
    "\n",
    "    coverage_dict = {}\n",
    "\n",
    "    for read in alignmentFile.fetch(until_eof=True):\n",
    "\n",
    "        nh = read.get_tag('NH')\n",
    "        inc = 1.0 / nh\n",
    "        ch = read.reference_name\n",
    "        start = read.reference_start\n",
    "        end = read.reference_end\n",
    "\n",
    "        update_coverage_dict(coverage_dict, ch, start, end, inc)\n",
    "        \n",
    "    return coverage_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bed_coverage_dict(in_bed):\n",
    "    \n",
    "    cov_dict = {}\n",
    "    with open(in_total) as itc:\n",
    "        for line in itc:\n",
    "            ls = line.split('\\t')\n",
    "            ch = ls[0].strip(' ')\n",
    "            pos = ls[1].strip(' ')\n",
    "            val = ls[2].strip(' \\n')\n",
    "            key = ch + ':' + pos\n",
    "            cov_dict[key] = val\n",
    "            \n",
    "    return cov_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step1 : create the scaffold coverage file with bedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in os.listdir(in_dir):\n",
    "    \n",
    "    if file.endswith('.bam'):\n",
    "        \n",
    "        base_name = file.split('.')[0]\n",
    "        in_path = os.path.join(in_dir, file)\n",
    "        out_path = os.path.join(out_dir, base_name + '.bed')\n",
    "        ! bedtools genomecov -d -ibam $in_path > $out_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2 : calculate coverages for bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/63_S2ChIPp.ztr.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIPp.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/63_S2ChIP_2.ztr.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP_2.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/65_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/65_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/80_S2ChIP_2.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP_2.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/80_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/80_S2Ph_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2Ph_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/80S2ChIP_1.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/80S2ChIP_1.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/301_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/302_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/324_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/491_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/504S2ChIP_1.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_1.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/504S2ChIP_2.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_2.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/530ChIP_1.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/530ChIP_1.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/530S2ChIP_2.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/530S2ChIP_2.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/591_S2PChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/591_S2PChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/638ChIP_1.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/638ChIP_1.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/63_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/80_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/491_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/324_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/302_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/301_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/638_S2_ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/638_S2_ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/1168_S2ChIP_1.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP_1.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/1168_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP.coverage.txt.\n",
      "bam file : /data/parastou/RNAdeg/results/RipChip/bams/510_S2ChIP.Aligned.sortedByCoord.out.bam ...\n",
      "Coverage file save in /data/parastou/RNAdeg/CoveragePlots/ChIP/510_S2ChIP.coverage.txt.\n"
     ]
    }
   ],
   "source": [
    "for file in os.listdir(in_dir):\n",
    "    \n",
    "    if file.endswith('.bam'):\n",
    "        \n",
    "        base_name = file.split('.')[0]\n",
    "        in_path = os.path.join(in_dir, file)\n",
    "        bed_path = os.path.join(out_dir, base_name + '.bed')\n",
    "        out_path = os.path.join(out_dir, base_name + '.coverage.txt')\n",
    "        \n",
    "        print('bam file : %s ...' % in_path)\n",
    "        ## Load a bam file and create coverage information for mapped regions\n",
    "        st = pysam.AlignmentFile(open(in_path, 'rb'))      \n",
    "        coverage_dict = generate_coverage_dict(st)\n",
    "        \n",
    "        ## Load bedtools coverage file as scafold, create coverage dictionary\n",
    "        bed_cov_dict = bed_coverage_dict(bed_path)\n",
    "        \n",
    "        assert len(bed_cov_dict) >= len(coverage_dict)\n",
    "        \n",
    "        ## Merge coverage information into a new file\n",
    "        with open(out_path, 'w+') as out:\n",
    "    \n",
    "            for coord, item in bed_cov_dict.items():\n",
    "                coverage = item\n",
    "                if coord in coverage_dict:\n",
    "                    coverage = coverage_dict[coord]\n",
    "                ch = coord.split(':')[0]\n",
    "                pos = coord.split(':')[1]\n",
    "                line = ch + '\\t' + str(pos) + '\\t' + str(coverage) + '\\n'\n",
    "                out.write(line)\n",
    "        print('Coverage file save in %s.' % out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3 : Generate Read-per-million normalized coverages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP_1.coverage.txt\n",
      "12140573.19616848\n",
      "12.14057319616848\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP_1.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP.coverage.txt\n",
      "4811017.73666634\n",
      "4.81101773666634\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/1168_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2_ChIP.coverage.txt\n",
      "4286852.040619181\n",
      "4.286852040619181\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2ChIP.coverage.txt\n",
      "5087370.131110993\n",
      "5.087370131110993\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/301_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2_ChIP.coverage.txt\n",
      "4968244.299999734\n",
      "4.968244299999735\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2ChIP.coverage.txt\n",
      "6028582.592165854\n",
      "6.028582592165854\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/302_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2_ChIP.coverage.txt\n",
      "3842185.43483367\n",
      "3.84218543483367\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2ChIP.coverage.txt\n",
      "6154374.983333238\n",
      "6.154374983333239\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/324_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2_ChIP.coverage.txt\n",
      "4216051.573000233\n",
      "4.216051573000232\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2ChIP.coverage.txt\n",
      "6679169.815665572\n",
      "6.679169815665572\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/491_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_1.coverage.txt\n",
      "4129087.0674050087\n",
      "4.129087067405009\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_1.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_2.coverage.txt\n",
      "4504533.804166626\n",
      "4.504533804166626\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/504S2ChIP_2.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/510_S2ChIP.coverage.txt\n",
      "13207270.82400024\n",
      "13.20727082400024\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/510_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/530ChIP_1.coverage.txt\n",
      "6246501.547904496\n",
      "6.246501547904496\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/530ChIP_1.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/530S2ChIP_2.coverage.txt\n",
      "3372411.0641190796\n",
      "3.3724110641190794\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/530S2ChIP_2.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/591_S2PChIP.coverage.txt\n",
      "9783825.272618288\n",
      "9.783825272618289\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/591_S2PChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/638ChIP_1.coverage.txt\n",
      "7909602.518236961\n",
      "7.909602518236961\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/638ChIP_1.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/638_S2_ChIP.coverage.txt\n",
      "6095531.489332845\n",
      "6.095531489332846\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/638_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP_2.coverage.txt\n",
      "5938591.960666602\n",
      "5.938591960666601\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP_2.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP.coverage.txt\n",
      "5940254.840666924\n",
      "5.940254840666924\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIPp.coverage.txt\n",
      "5938591.960666602\n",
      "5.938591960666601\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/63_S2ChIPp.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/65_S2ChIP.coverage.txt\n",
      "6280500.11166617\n",
      "6.28050011166617\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/65_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80S2ChIP_1.coverage.txt\n",
      "9360520.529665375\n",
      "9.360520529665376\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80S2ChIP_1.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP_2.coverage.txt\n",
      "6388519.298333078\n",
      "6.388519298333078\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP_2.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP.coverage.txt\n",
      "5377833.16499994\n",
      "5.377833164999941\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2_ChIP.coverage.txt\n",
      "6388519.298333078\n",
      "6.388519298333078\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2_ChIP.norm.coverage.txt\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2Ph_ChIP.coverage.txt\n",
      "2639820.889999614\n",
      "2.6398208899996143\n",
      "/data/parastou/RNAdeg/CoveragePlots/ChIP/80_S2Ph_ChIP.norm.coverage.txt\n"
     ]
    }
   ],
   "source": [
    "for file in os.listdir(out_dir):\n",
    "    \n",
    "    if file.endswith('.coverage.txt'):\n",
    "        \n",
    "        in_path = os.path.join(out_dir, file)\n",
    "        print(in_path)\n",
    "        base_name = file.split('.')[0]\n",
    "        out_path = os.path.join(out_dir, base_name + '.norm.coverage.txt')\n",
    "        \n",
    "        df = df = pd.read_csv(in_path, sep='\\t', names=range(3))\n",
    "        total = sum(df[2]) / 50\n",
    "        print(total)\n",
    "        scaling_factor = (total / 1000000)\n",
    "        print(scaling_factor)\n",
    "        df[2] = round(df[2] / scaling_factor, 2)\n",
    "        df.to_csv(out_path, sep='\\t', index=None, header=None)\n",
    "        print(out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sanity check for library depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63_S2ChIPp\n",
      "5.356333999892295\n",
      "63_S2ChIP_2\n",
      "5.356333999892295\n",
      "65_S2ChIP\n",
      "6.267980999970094\n",
      "80_S2ChIP_2\n",
      "6.328341999773746\n",
      "80_S2ChIP\n",
      "5.357875999907185\n",
      "80_S2Ph_ChIP\n",
      "1.7535309999972273\n",
      "80S2ChIP_1\n",
      "9.338956999841269\n",
      "301_S2ChIP\n",
      "5.060596999987656\n",
      "302_S2ChIP\n",
      "6.005917999931659\n",
      "324_S2ChIP\n",
      "6.132901999953194\n",
      "491_S2ChIP\n",
      "6.6667069999678406\n",
      "504S2ChIP_1\n",
      "4.112042000080235\n",
      "504S2ChIP_2\n",
      "4.489328000043439\n",
      "530ChIP_1\n",
      "6.190077999752234\n",
      "530S2ChIP_2\n",
      "3.334843000082868\n",
      "591_S2PChIP\n",
      "9.78366099994927\n",
      "638ChIP_1\n",
      "7.90067499979194\n",
      "63_S2ChIP\n",
      "5.344684999903266\n",
      "80_S2_ChIP\n",
      "6.328341999773746\n",
      "491_S2_ChIP\n",
      "4.098884000109946\n",
      "324_S2_ChIP\n",
      "3.8086730001036035\n",
      "302_S2_ChIP\n",
      "4.960510000015301\n",
      "301_S2_ChIP\n",
      "4.256788000142784\n",
      "638_S2_ChIP\n",
      "6.007343999866427\n",
      "1168_S2ChIP_1\n",
      "12.099033000172284\n",
      "1168_S2ChIP\n",
      "2.5741050000027546\n",
      "510_S2ChIP\n",
      "13.202423000104327\n"
     ]
    }
   ],
   "source": [
    "for file in os.listdir(in_dir):\n",
    "    \n",
    "    if file.endswith('.bam'):\n",
    "        \n",
    "        base_name = file.split('.')[0]\n",
    "        in_path = os.path.join(in_dir, file)       \n",
    "        print(base_name)\n",
    "        \n",
    "        ## Load a bam file and create coverage information for mapped regions\n",
    "        st = pysam.AlignmentFile(open(in_path, 'rb'))\n",
    "        total = 0\n",
    "        for r in st.fetch(until_eof=True):\n",
    "            nh = r.get_tag('NH')\n",
    "            inc = 1.0 / nh\n",
    "            total += inc\n",
    "        print(total/1000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
